overlord/fdestate/handlers.go

```diff
@@ -212,7 +215,146 @@ func (m *FDEManager) doRenameKeys(t *state.Task, tomb *tomb.Tomb) error {
 }
 
 func (m *FDEManager) doAddProtectedKeys(t *state.Task, _ *tomb.Tomb) (err error) {
-	// TODO:FDEM: implement handler
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	var keyslotRefs []KeyslotRef
+	if err := t.Get("keyslots", &keyslotRefs); err != nil {
+		return err
+	}
+
+	var authMode device.AuthMode
+	if err := t.Get("auth-mode", &authMode); err != nil {
+		return err
+	}
+	switch authMode {
+	case device.AuthModeNone, device.AuthModePassphrase, device.AuthModePIN:
+	default:
+		return fmt.Errorf("internal error: unexpected authentication mode %q", authMode)
+	}
+
+	var keyslotRoles map[string][]string
+	if err := t.Get("roles", &keyslotRoles); err != nil {
+		return err
+	}
+
+	var volumesAuth *device.VolumesAuthOptions
+	// authentication options are only relevant for PINs and passphrases.
+	switch authMode {
+	case device.AuthModePassphrase, device.AuthModePIN:
+		cached := m.state.Cached(volumesAuthOptionsKey{})
+		if cached == nil {
+			return errors.New("cannot find authentication options in memory: unexpected snapd restart")
+		}
+		var ok bool
+		volumesAuth, ok = cached.(*device.VolumesAuthOptions)
+		if !ok {
+			return fmt.Errorf("internal error: wrong data type under volumesAuthOptionsKey: %T", cached)
+		}
+		if err := volumesAuth.Validate(); err != nil {
+			return fmt.Errorf("internal error: invalid authentication options: %v", err)
+		}
+	}
+
+	// XXX: unlock state and let conflict detection handle the rest?
```

Comment:
do we still need this XXX?

Reply:
I am keeping it as a reminder so that if we face some serious slowdown/hang we could explore unlocking the state.

To be safe, I rather have the state locked until we have a good reason not to.